Optimized version of SymIntArrayRef

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86235
* __->__ #86233
* #86230

Previously, to do a conversion of SymIntArrayRef to IntArrayRef, you
always had to a loop over the elements checking if any of them were
symbolic.  Now, at the cost of an extra boolean field on SymIntArrayRef,
we have a bool field you can check directly to see if any element
of SymIntArrayRef is symbolic.  This optimization pays off for eager mode,
as our dispatcher typically deals in SymIntArrayRef and then converts
to IntArrayRef when we call into the C++ kernels.

The cost is in code: I have to duplicate the ArrayRef implementation for
SymInt and do a bunch of modifications.  There is a lot of code and
unfortunately you have to review it all. The main changes:

- Tracking and propagating of AnySymbolic
- A lot of constexpr removal

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Optimized version of SymIntArrayRef

Previously, to do a conversion of SymIntArrayRef to IntArrayRef, you
always had to a loop over the elements checking if any of them were
symbolic.  Now, at the cost of an extra boolean field on SymIntArrayRef,
we have a bool field you can check directly to see if any element
of SymIntArrayRef is symbolic.  This optimization pays off for eager mode,
as our dispatcher typically deals in SymIntArrayRef and then converts
to IntArrayRef when we call into the C++ kernels.

The cost is in code: I have to duplicate the ArrayRef implementation for
SymInt and do a bunch of modifications.  There is a lot of code and
unfortunately you have to review it all. The main changes:

- Tracking and propagating of AnySymbolic
- A lot of constexpr removal

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Optimized version of SymIntArrayRef"


Previously, to do a conversion of SymIntArrayRef to IntArrayRef, you
always had to a loop over the elements checking if any of them were
symbolic.  Now, at the cost of an extra boolean field on SymIntArrayRef,
we have a bool field you can check directly to see if any element
of SymIntArrayRef is symbolic.  This optimization pays off for eager mode,
as our dispatcher typically deals in SymIntArrayRef and then converts
to IntArrayRef when we call into the C++ kernels.

The cost is in code: I have to duplicate the ArrayRef implementation for
SymInt and do a bunch of modifications.  There is a lot of code and
unfortunately you have to review it all. The main changes:

- Tracking and propagating of AnySymbolic
- A lot of constexpr removal

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Optimized version of SymIntArrayRef"


Previously, to do a conversion of SymIntArrayRef to IntArrayRef, you
always had to a loop over the elements checking if any of them were
symbolic.  Now, at the cost of an extra boolean field on SymIntArrayRef,
we have a bool field you can check directly to see if any element
of SymIntArrayRef is symbolic.  This optimization pays off for eager mode,
as our dispatcher typically deals in SymIntArrayRef and then converts
to IntArrayRef when we call into the C++ kernels.

The cost is in code: I have to duplicate the ArrayRef implementation for
SymInt and do a bunch of modifications.  There is a lot of code and
unfortunately you have to review it all. The main changes:

- Tracking and propagating of AnySymbolic
- A lot of constexpr removal

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Optimized version of SymIntArrayRef"


Previously, to do a conversion of SymIntArrayRef to IntArrayRef, you
always had to a loop over the elements checking if any of them were
symbolic.  Now, at the cost of an extra boolean field on SymIntArrayRef,
we have a bool field you can check directly to see if any element
of SymIntArrayRef is symbolic.  This optimization pays off for eager mode,
as our dispatcher typically deals in SymIntArrayRef and then converts
to IntArrayRef when we call into the C++ kernels.

The cost is in code: I have to duplicate the ArrayRef implementation for
SymInt and do a bunch of modifications.  There is a lot of code and
unfortunately you have to review it all. The main changes:

- Tracking and propagating of AnySymbolic
- A lot of constexpr removal

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Optimized version of SymIntArrayRef"


Previously, to do a conversion of SymIntArrayRef to IntArrayRef, you
always had to a loop over the elements checking if any of them were
symbolic.  Now, at the cost of an extra boolean field on SymIntArrayRef,
we have a bool field you can check directly to see if any element
of SymIntArrayRef is symbolic.  This optimization pays off for eager mode,
as our dispatcher typically deals in SymIntArrayRef and then converts
to IntArrayRef when we call into the C++ kernels.

The cost is in code: I have to duplicate the ArrayRef implementation for
SymInt and do a bunch of modifications.  There is a lot of code and
unfortunately you have to review it all. The main changes:

- Tracking and propagating of AnySymbolic
- A lot of constexpr removal

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Optimized version of SymIntArrayRef"


Previously, to do a conversion of SymIntArrayRef to IntArrayRef, you
always had to a loop over the elements checking if any of them were
symbolic.  Now, at the cost of an extra boolean field on SymIntArrayRef,
we have a bool field you can check directly to see if any element
of SymIntArrayRef is symbolic.  This optimization pays off for eager mode,
as our dispatcher typically deals in SymIntArrayRef and then converts
to IntArrayRef when we call into the C++ kernels.

The cost is in code: I have to duplicate the ArrayRef implementation for
SymInt and do a bunch of modifications.  There is a lot of code and
unfortunately you have to review it all. The main changes:

- Tracking and propagating of AnySymbolic
- A lot of constexpr removal

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Optimized version of SymIntArrayRef"


Previously, to do a conversion of SymIntArrayRef to IntArrayRef, you
always had to a loop over the elements checking if any of them were
symbolic.  Now, at the cost of an extra boolean field on SymIntArrayRef,
we have a bool field you can check directly to see if any element
of SymIntArrayRef is symbolic.  This optimization pays off for eager mode,
as our dispatcher typically deals in SymIntArrayRef and then converts
to IntArrayRef when we call into the C++ kernels.

The cost is in code: I have to duplicate the ArrayRef implementation for
SymInt and do a bunch of modifications.  There is a lot of code and
unfortunately you have to review it all. The main changes:

- Tracking and propagating of AnySymbolic
- A lot of constexpr removal

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD commented on c10/core/SymIntArrayRef.h:
```diff
@@ -13,6 +13,337 @@
+        "created ArrayRef with nullptr and non-zero length! c10::optional relies on this being illegal");
+  }
+
+  void debugCheckAnySymbolicInvariant() {
```
nit: should this and the one above be marked inline? Or there is no way the compiler will miss these?

ezyang replied:
Since they compile to empty bodies sans DEBUG I think it's safe to say the compiler will get it

albanD commented on c10/core/SymIntArrayRef.h:
```diff
@@ -13,6 +13,337 @@
+  ArrayRef(Unchecked, const int64_t* data, size_t length)
+      : Data(reinterpret_cast<const c10::SymInt*>(data)),
+        Length(length),
+        AnySymbolic(false) {
```
Should we have an unchecked version for symbolic=true as well? Or we just don't need that today?

albanD commented on c10/core/SymIntArrayRef.h:
```diff
@@ -13,6 +13,337 @@
+  }
+
+  /// Construct an ArrayRef from a std::vector.
+  // The enable_if stuff here makes sure that this isn't used for
```
The enable_if logic is not here anymore. It is actually a static assert

albanD commented on c10/core/SymIntArrayRef.h:
```diff
@@ -13,6 +13,337 @@
+  /* implicit */ ArrayRef(const std::vector<T, A>& Vec)
+      : Data(Vec.data()), Length(Vec.size()), AnySymbolic(computeSymbolic()) {
+    static_assert(
+        !std::is_same<T, bool>::value,
```
Should we forbid anything that is not int ~~or float~~ (in all the constructors actually)?

ezyang replied:
This is dead actually, should delete

albanD commented on c10/core/SymIntArrayRef.h:
```diff
@@ -13,6 +13,337 @@
+  /// Construct an ArrayRef from a std::array
+  template <size_t N>
+  /* implicit */ ArrayRef(const std::array<T, N>& Arr)
+      : Data(Arr.data()), Length(N), AnySymbolic(computeSymbolic()) {}
```
Ho, std::vector<bool> is bad but std::array<bool> is ok?

ezyang replied:
yes, because array of bool is not specialized to be bitfield (not really relevant here, but that's why)

albanD commented on c10/core/SymIntArrayRef.h:
```diff
@@ -13,6 +13,337 @@
+    return Data[Index];
+  }
+
+  /// Disallow accidental assignment from a temporary.
```
Are the other assignments the default? We should override them to update the AnySymbolic field right?

ezyang replied:
The assignments will copy over all the fields, which is appropriate here.

albanD replied:
I was talking about single item assignment `array[i] = some_symint`.

ezyang replied:
this is indeed a problem, and I need a much more complicated pr to overcome it

albanD commented on c10/core/SymIntArrayRef.h:
```diff
@@ -13,6 +13,337 @@
+};
+
+template <>
+class arrayref_optional_base<ArrayRef<SymInt>> {
```
Why do we need this one explicitly? I don't see any difference with the other one?

ezyang replied:
There's an extra field in uninitialized, which is necessary to make the memcpy sizes line up

albanD reviewed (commented):
Sounds good! Just a lot of questions

ezyang committed:
Update on "Optimized version of SymIntArrayRef"


Previously, to do a conversion of SymIntArrayRef to IntArrayRef, you
always had to a loop over the elements checking if any of them were
symbolic.  Now, at the cost of an extra boolean field on SymIntArrayRef,
we have a bool field you can check directly to see if any element
of SymIntArrayRef is symbolic.  This optimization pays off for eager mode,
as our dispatcher typically deals in SymIntArrayRef and then converts
to IntArrayRef when we call into the C++ kernels.

The cost is in code: I have to duplicate the ArrayRef implementation for
SymInt and do a bunch of modifications.  There is a lot of code and
unfortunately you have to review it all. The main changes:

- Tracking and propagating of AnySymbolic
- A lot of constexpr removal

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented on c10/core/SymIntArrayRef.cpp:
```diff
@@ -13,16 +13,13 @@ at::IntArrayRef asIntArrayRefSlow(c10::SymIntArrayRef ar) {
 }
 
 c10::optional<at::IntArrayRef> asIntArrayRefSlowOpt(c10::SymIntArrayRef ar) {
```
Well, it does do a boolean test, but yes, it's not that slow. I'll whack this in a follow up

ezyang commented:
I don't think we can actually land this without also adding specializations for the owning container types that hold SymIntArrayRef

ezyang commented:
not enough benefit for the work
